### PR TITLE
[docker] Fix restart behavior with Docker

### LINKED ISSUE #12104: [autoscaler] [docker] ray up on stopped node fails?
### What is the problem?

I ran `ray up` on a cluster that I had previously stopped. 

The ray up was unable to finish properly due to a `tmp` call.

cc @ijrsvt 
```
(base) ➜  aws git:(docker-fix) ✗ ray up example-full.yaml -y
Cluster: default

Loaded cached provider configuration
If you experience issues with the cloud provider, try re-running the command with --no-config-cache.
AWS config
  IAM Profile: ray-autoscaler-v1 [default]
  EC2 Key pair (head & workers): ray-autoscaler_2_us-west-2 [default]
  VPC Subnets (head & workers): subnet-d26e76b6, subnet-503c0e26 [default]
  EC2 Security groups (head & workers): sg-0e2ad6debfc8e0814 [default]
  EC2 AMI (head & workers): ami-0a2363a9cff180a64

No head node found. Launching a new cluster. Confirm [y/N]: y [automatic, due to --yes]

Acquiring an up-to-date head node
  Reusing nodes i-0c90dd53359e25223. To disable reuse, set `cache_stopped_nodes: False` under `provider` in the cluster configuration.
  Stopping instances to reuse
  Launched a new head node
  Fetching the new head node

<1/1> Setting up head node
  Prepared bootstrap config
  New status: waiting-for-ssh
  [1/6] Waiting for SSH to become available
    Running `uptime` as a test.
    Waiting for IP
      Not yet available, retrying in 10 seconds
      Received: 35.165.135.138
ssh: connect to host 35.165.135.138 port 22: Connection refused
    SSH still not available SSH command failed., retrying in 5 seconds.
Warning: Permanently added '35.165.135.138' (ECDSA) to the list of known hosts.
 09:00:19 up 0 min,  1 user,  load average: 0.35, 0.08, 0.03
Shared connection to 35.165.135.138 closed.
    Success.
Shared connection to 35.165.135.138 closed.
latest-gpu: Pulling from rayproject/ray
Digest: sha256:9e330168fbeface86427d29b4a1a996bdefca42fff38ff155a29c8d1d1020b74
Status: Image is up to date for rayproject/ray:latest-gpu
docker.io/rayproject/ray:latest-gpu
Shared connection to 35.165.135.138 closed.
Shared connection to 35.165.135.138 closed.
Shared connection to 35.165.135.138 closed.
Shared connection to 35.165.135.138 closed.
NVIDIA-SMI has failed because it couldn't communicate with the NVIDIA driver. Make sure that the latest NVIDIA driver is installed and running.

Shared connection to 35.165.135.138 closed.
2020-11-18 01:00:24,949	WARNING command_runner.py:780 -- Nvidia Container Runtime is present, but no GPUs found.
Shared connection to 35.165.135.138 closed.
9191520da0ddb9d6259795e5a758a298934bdab25ee153e60ddcd350318ed789
Shared connection to 35.165.135.138 closed.
Shared connection to 35.165.135.138 closed.
lstat /tmp/ray_tmp_mount: no such file or directory
Shared connection to 35.165.135.138 closed.
  New status: update-failed
  !!!
  SSH command failed.
  !!!

  Failed to setup head node.
```

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?

The Docker command runner does not handle restarting from a stopped node correctly. This is because (from the perspective of the container) a stopped node (power down) and a terminated node (deleting the hard disk) is the same. 

## Related issue number

Closes #12104

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/command_runner.py [hunk 1]
```diff
@@ -75,11 +75,14 @@ def remote_shell_command_str(self) -> str:
...
-    def run_init(self, *, as_head: bool, file_mounts: Dict[str, str]) -> None:
+    def run_init(self, *, as_head: bool,
+                 file_mounts: Dict[str, str]) -> Optional[bool]:
         """Used to run extra initialization commands.
 
         Args:
             as_head (bool): Run as head image or worker.
             file_mounts (dict): Files to copy to the head and worker nodes.
+        Returns:
+            optional (bool): Whether initialization was run.
```
Comment (ijrsvt) [inline python/ray/autoscaler/command_runner.py:86]: @AmeerHajAli This is an early draft of a PR to allow changing file mounts with Docker & also to help fix stop (but not terminate) with Docker. Before I go any farther with this approach, I wanted to see if this change in CommandRunner is acceptable?

Comment (AmeerHajAli) [reply python/ray/autoscaler/command_runner.py:86]: Yes, bit you need to modify the function here too:
https://github.com/anyscale/product/blob/8462c65ff0e3a2f04e915e3d914e14634cdbae3c/backend/server/services/gateway_commandrunner_service.py#L86
1) edit the type annotation.
2) make it return `self.provider.get_response(request)`

**Comment by richardliaw** [comment]
merge broken? maybe need to open new PR?

**Comment by ijrsvt** [comment]
> merge broken? maybe need to open new PR?

Yikes yep!

**Pull request closed without merging**